xds/internal/xdsclient/tlscreds/bundle_test.go

```diff
@@ -0,0 +1,256 @@
+	}
+	// close the server and create a new one to force client to do a new
+	// handshake.
+	stopSrv()
+
+	invalidCa, err := os.ReadFile(testdata.Path("ca.pem"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %v", err)
+	}
+	// unload root cert
+	err = os.WriteFile(caPath, invalidCa, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test CA cert: %v", err)
+	}
+
+	// Leave time for the file_watcher provider to reload the CA.
+	time.Sleep(100 * time.Millisecond)
+
+	_, stopFunc := startServer(t, srvAddr.String(), tls.NoClientCert)
+	defer stopFunc()
+
+	// Client handshake should fail because the server cert is signed by an
+	// unknown CA.
+	_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+	if st, ok := status.FromError(err); !ok || st.Code() != codes.Unavailable {
+		t.Errorf("Expected unavailable error, got %v", err)
+	} else if want := "certificate signed by unknown authority"; !strings.Contains(st.Message(), want) {
+		t.Errorf("Expected call error to contain '%s', got %v", want, st.Message())
+	}
+}
+
+func TestMTLS(t *testing.T) {
+	srvAddr, stopFunc := startServer(t, "localhost:0", tls.RequireAndVerifyClientCert)
+	defer stopFunc()
+
+	cfg := fmt.Sprintf(`{
+		"ca_certificate_file": "%s",
+		"certificate_file": "%s",
+		"private_key_file": "%s"
+	}`,
+		testdata.Path("x509/server_ca_cert.pem"),
+		testdata.Path("x509/client1_cert.pem"),
+		testdata.Path("x509/client1_key.pem"))
+	tlsBundle, err := NewBundle([]byte(cfg))
+	if err != nil {
+		t.Fatalf("Failed to create TLS bundle: %v", err)
+	}
+	dialOpts := []grpc.DialOption{
+		grpc.WithCredentialsBundle(tlsBundle),
+		grpc.WithAuthority("x.test.example.com"),
+	}
+
+	t.Run("ValidClientCert", func(t *testing.T) {
+		conn, err := grpc.Dial(srvAddr.String(), dialOpts...)
+		if err != nil {
+			t.Fatalf("Error dialing: %v", err)
+		}
+		client := testgrpc.NewTestServiceClient(conn)
+
+		_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+		if err != nil {
+			t.Errorf("Error calling EmptyCall: %v", err)
+		}
+		conn.Close()
+	})
+
+	t.Run("Provider failing", func(t *testing.T) {
+		// Check that if the provider returns an errors, we fail the handshake.
+		// It's not easy to trigger this condition, so we rely on closing the
+		// provider.
+		creds, ok := tlsBundle.TransportCredentials().(*reloadingCreds)
+
+		// Force the provider to be initialized. The test is flaky otherwise,
+		// since close may be a noop.
+		_, _ = creds.provider.KeyMaterial(context.Background())
+
+		if !ok {
+			t.Fatalf("Expected reloadingCreds, got %T", tlsBundle.TransportCredentials())
+		}
+
+		creds.provider.Close()
+
+		conn, err := grpc.Dial(srvAddr.String(), dialOpts...)
+		if err != nil {
+			t.Fatalf("Error dialing: %v", err)
+		}
+		client := testgrpc.NewTestServiceClient(conn)
+		_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+		if st, ok := status.FromError(err); !ok || st.Code() != codes.Unavailable {
```

Comment:
There is a `status.Code()` function that returns the status code given an error.

Reply:
And I think you could simply do `err.Error()` and compare it against `wantErr` instead of getting the status from the error and getting the message.

Reply:
Done, but I used `strings.HasSuffix` to make sure the wanted error is at the end of err. `err.Error()` looks like:

```
rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: provider instance is closed
```